domain/application/state/storage.go

```diff
@@ -38,6 +38,80 @@ const (
 	storageNamespace    = domainsequence.StaticNamespace("storage")
 )
 
+// GetApplicationStorage returns the storage directives set for an application,
+// keyed to the storage name. If the application does not have any storage
+// directives set then an empty result is returned.
+//
+// If the application does not exist, then a [applicationerrors.ApplicationNotFound]
+// error is returned.
+func (st *State) GetApplicationStorage(
+	ctx context.Context,
+	appUUID coreapplication.UUID,
+) (application.ApplicationStorage, error) {
+	db, err := st.DB(ctx)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	appUUIDInput := entityUUID{UUID: appUUID.String()}
+
+	query, err := st.Prepare(`
+SELECT &applicationStorageInfo.* FROM (
+	SELECT  asd.count,
+			asd.size_mib,
+			asd.storage_name,
+			sp.name as storage_pool_name
+	FROM application_storage_directive asd
+	JOIN storage_pool sp ON sp.uuid = asd.storage_pool_uuid
+	WHERE asd.application_uuid = $entityUUID.uuid
+)
+		`,
+		appUUIDInput, applicationStorageInfo{},
+	)
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	storageInfoResult := []applicationStorageInfo{}
+
+	if err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		exists, err := st.checkApplicationExists(ctx, tx, appUUID)
+		if err != nil {
+			return errors.Errorf(
+				"checking application %q exists: %w", appUUID, err,
+			)
+		}
+		if !exists {
+			return errors.Errorf(
+				"application %q does not exist", appUUID,
+			).Add(applicationerrors.ApplicationNotFound)
+		}
+
+		err = tx.Query(ctx, query, appUUIDInput).GetAll(&storageInfoResult)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return err
+	}); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	// Preallocate map with expected size since each storage name
+	// should be unique within an application
+	// TODO: Although we do have the composite key on charm_uuid, so in theory if this was different
+	// this would break?
```

Comment:
This is fine. I do believe that the data type chosen for the return here is being driven by facade design which isn't needed. I personally would have returned this as a slice. Will leave it up to you.

+ remove TODO
+ remove "should" statement. It is fact and must be the case.